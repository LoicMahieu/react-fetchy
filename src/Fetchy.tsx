import { isEqual, pick } from "lodash";
import * as PropTypes from "prop-types";
import * as React from "react";
import * as request from "superagent";

export interface IState {
  fulfilled: boolean;
  pending: boolean;
  rejected: boolean;

  error: any;
  progress: any;
  result: any;
  value: any;
}

type methodType = "get" | "post" | "put" | "delete";

export interface IOptions {
  body?: any;
  method?: methodType;
  url?: string;
  query?: object;
  headers?: object;
  then?(state: IState): Promise<any>;
}

export interface IBag {
  state: IState;
  fetch(options: IOptions): Promise<IState>;
  reset(): void;
}

interface IProps extends IOptions {
  children?(bag: IBag): React.ReactNode;
  render?(bag: IBag): React.ReactNode;
}

const initialState: IState = {
  fulfilled: false,
  pending: false,
  rejected: false,

  error: null,
  progress: null,
  result: null,
  value: null,
};

const compareProps = [
  "url",
  "query",
  "headers",
  "method",
];

export default class Fetchy extends React.Component<IProps, IState> {
  public static defaultProps = {};

  public static propType = {
    title: PropTypes.string,
  };

  public readonly state: IState = initialState;

  private req: any;

  public async componentDidMount() {
    if (this.props.url) {
      this.fetch();
    }
  }

  public componentDidUpdate(prevProps: IProps) {
    if (!isEqual(pick(this.props, compareProps), pick(prevProps, compareProps))) {
      this.fetch();
    }
  }

  public componentWillUnmount() {
    if (this.req) {
      this.req.abort();
    }
  }

  public render() {
    const render = typeof this.props.children === "function"
      ? this.props.children
      : this.props.render;

    if (!render) {
      throw new Error("Missing children or render props");
    }

    const bag: IBag = {
      fetch: this.fetch,
      reset: this.reset,
      state: Object.freeze({ ...this.state }),
    };

    return render(bag);
  }

  private fetch = async (options: IOptions = {}) => {
    if (this.req) {
      this.req.abort();
    }

    const mergedOptions = {
      ...this.props,
      ...options,
    };

    const {
      url,
      method = "get",
      query,
      body,
      headers,
      then,
    } = mergedOptions;

    if (!url) {
      throw new Error("Parameter `url` is required.");
    }

    this.setState({ pending: true, fulfilled: false });

    let response: any;

    // Create request
    const m = method.toLowerCase() as methodType;
    const createReq = request[m];
    if (!createReq)Â {
      throw new Error(`Invalid method ${method}.`);
    }
    const req = createReq(url);
    this.req = req;

    if (headers) {
      req.set(headers);
    }
    if (query) {
      req.query(query);
    }
    if (body) {
      req.send(body);
    }
    req.on("progress", (progress) => {
      this.setState((oldState) => ({ ...oldState, progress }));
    });
    req.on("response", (incomingMessage) => {
      response = incomingMessage;
    });

    try {
      // Wait response
      await req;

      /* istanbul ignore if : should never happend */
      if (!response) {
        throw new Error("No response");
      }

      // Handle Result
      const value = response.body;
      const state: IState = {
        ...initialState,
        fulfilled: true,
        result: response,
        value,
      };
      if (then) {
        Object.assign(state, then(state));
      }
      this.setState(state);
    } catch (error) {
      error.data = error.data || {};
      error.data.result = {
        response,
      };

      this.setState({
        error,
        fulfilled: false,
        pending: false,
        rejected: true,
        result: response,
        value: null,
      });

      throw error;
    }

    return Object.freeze({ ...this.state });
  }

  private reset = () => {
    this.setState({ ...initialState });
  }
}
