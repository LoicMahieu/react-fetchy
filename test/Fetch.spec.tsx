
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { readFileSync } from "fs";
import { createServer } from "http";
import { join as pathJoin } from "path";
import * as React from "react";

import Fetchy, { IBag } from "../src";

const base = "http://localhost:8080";

configure({ adapter: new Adapter() });

describe("Fetchy", () => {
  let server;
  beforeAll((done) => {
    server = createServer((req, res) => {
      res.setHeader("Date", "Fake");

      switch (true) {
        case /^\/200/.test(req.url):
          res.setHeader("Content-Type", "application/json");
          res.write(JSON.stringify({
            headers: req.headers,
            url: req.url,
          }));
          res.end();
          break;
        case /^\/404/.test(req.url):
          res.statusCode = 404;
          res.write("404");
          res.end();
          break;
        case /^\/post\/200/.test(req.url):
          let body = "";
          req.on("data", (data) => {
            body += data;
          });
          req.on("end", () => {
            res.write(JSON.stringify({
              body,
              headers: req.headers,
              url: req.url,
            }));
            res.end();
          });
          break;
        default:
          res.write("ok");
          res.end();
          break;
      }
    });

    server.listen(8080, done);
  });
  afterAll(() => {
    if (server) {
      server.close();
    }
  });

  it("no render or children", () => {
    expect(() => shallow(<Fetchy />)).toThrow("Missing children or render props");
  });
  describe("basic", () => {
    it("render", () => {
      const render = jest.fn((bag: IBag) => {
        expect(bag).toMatchSnapshot();
      });
      shallow(<Fetchy render={render} />);
      expect(render).toHaveBeenCalled();
    });
    it("children", () => {
      const children = jest.fn((bag: IBag) => {
        expect(bag).toMatchSnapshot();
      });
      shallow(<Fetchy children={children} />);
      expect(children).toHaveBeenCalled();
    });
  });
  describe("fetch", () => {
    let bag;
    let el;
    beforeEach(() => {
      const render = jest.fn((iBag: IBag) => {
        bag = iBag;
      });
      el = <Fetchy render={render} />;
      shallow(el);
      expect(render).toHaveBeenCalled();
      expect(bag).toMatchSnapshot();
    });

    it("throw error if no url", async () => {
      expect.assertions(3);
      try {
        await bag.fetch();
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });

    it("GET 200", async () => {
      expect(await bag.fetch({
        headers: {
          "Some-Header": "1",
        },
        query: "foo",
        retry: 5,
        then: (value) => ({
          ...value,
          newValue: 1,
        }),
        timeout: 20,
        url: `${base}/200`,
        withCredentials: true,
      })).toMatchSnapshot();
      expect(bag).toMatchSnapshot();
    });

    it("GET 404", async () => {
      expect.assertions(5);
      try {
        await bag.fetch({ url: `${base}/404` });
      } catch (e) {
        expect(e.data).toMatchSnapshot();
        expect(e).toMatchSnapshot();
      }
      expect(bag).toMatchSnapshot();
    });

    it("POST 200", async () => {
      expect(await bag.fetch({
        body: Buffer.from("hello world"),
        method: "post",
        url: `${base}/post/200`,
      })).toMatchSnapshot();
      expect(bag).toMatchSnapshot();
    });

    it("abort", async () => {
      const promise = bag.fetch({ url: `${base}/200?first` });
      expect(bag).toMatchSnapshot();
      await bag.fetch({ url: `${base}/200?second` });
      expect(bag).toMatchSnapshot();
    });

    it("invalid method", async () => {
      try {
        await bag.fetch({ url: `${base}`, method: "FOO" });
      } catch (error) {
        expect(error.message).toEqual("Invalid method FOO.");
      }
    });

    it("accept method with case", async () => {
      expect(await bag.fetch({
        method: "GET",
        url: `${base}/200`,
      })).toMatchSnapshot();
    });
  });
  describe("reset", () => {
    it("works", async () => {
      let bag;
      const render = jest.fn((iBag: IBag) => {
        bag = iBag;
      });
      const element = <Fetchy render={render} />;
      shallow(element);
      expect(bag).toMatchSnapshot();
      await bag.fetch({ url: `${base}/200` });
      expect(bag).toMatchSnapshot();
      await bag.reset();
      expect(bag).toMatchSnapshot();
    });
  });
  it("fetch on mount / unmount", async () => {
    let bag;
    const render = jest.fn((iBag: IBag) => {
      bag = iBag;
    });
    const element = <Fetchy render={render} url={`${base}/200`} />;
    const wrapper = shallow(element);
    expect(bag.state.pending).toEqual(true);
    wrapper.unmount();
  });
  it("unmount without any fetch", async () => {
    let bag;
    const render = jest.fn((iBag: IBag) => {
      bag = iBag;
    });
    const element = <Fetchy render={render} />;
    const wrapper = shallow(element);
    expect(bag.state.pending).toEqual(false);
    wrapper.unmount();
  });
  it("fetch on props change", async () => {
    let bag;
    const render = jest.fn((iBag: IBag) => {
      bag = iBag;
    });
    const wrapper = shallow(<Fetchy render={render} />);
    expect(bag.state.pending).toEqual(false);
    wrapper.setProps({ url: `${base}/200` });
    expect(bag.state.pending).toEqual(true);
    wrapper.unmount();
  });
});
