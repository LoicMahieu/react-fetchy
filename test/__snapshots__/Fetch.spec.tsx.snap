// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetchy basic children 1`] = `
Object {
  "fetch": [Function],
  "reset": [Function],
  "state": Object {
    "error": null,
    "fulfilled": false,
    "pending": false,
    "progress": null,
    "rejected": false,
    "result": null,
    "value": null,
  },
}
`;

exports[`Fetchy basic render 1`] = `
Object {
  "fetch": [Function],
  "reset": [Function],
  "state": Object {
    "error": null,
    "fulfilled": false,
    "pending": false,
    "progress": null,
    "rejected": false,
    "result": null,
    "value": null,
  },
}
`;

exports[`Fetchy fetch GET 200 1`] = `
Object {
  "fetch": [Function],
  "reset": [Function],
  "state": Object {
    "error": null,
    "fulfilled": false,
    "pending": false,
    "progress": null,
    "rejected": false,
    "result": null,
    "value": null,
  },
}
`;

exports[`Fetchy fetch GET 200 2`] = `
Object {
  "error": null,
  "fulfilled": true,
  "newValue": 1,
  "pending": false,
  "progress": null,
  "rejected": false,
  "result": Object {
    "header": Object {
      "connection": "close",
      "content-type": "application/json",
      "date": "Fake",
      "transfer-encoding": "chunked",
    },
    "req": Object {
      "data": undefined,
      "headers": Object {
        "some-header": "1",
        "user-agent": "node-superagent/3.8.3",
      },
      "method": "GET",
      "url": "http://localhost:8080/200?foo",
    },
    "status": 200,
    "text": "{\\"headers\\":{\\"host\\":\\"localhost:8080\\",\\"accept-encoding\\":\\"gzip, deflate\\",\\"user-agent\\":\\"node-superagent/3.8.3\\",\\"some-header\\":\\"1\\",\\"connection\\":\\"close\\"},\\"url\\":\\"/200?foo\\"}",
  },
  "value": Object {
    "headers": Object {
      "accept-encoding": "gzip, deflate",
      "connection": "close",
      "host": "localhost:8080",
      "some-header": "1",
      "user-agent": "node-superagent/3.8.3",
    },
    "url": "/200?foo",
  },
}
`;

exports[`Fetchy fetch GET 200 3`] = `
Object {
  "fetch": [Function],
  "reset": [Function],
  "state": Object {
    "error": null,
    "fulfilled": true,
    "newValue": 1,
    "pending": false,
    "progress": null,
    "rejected": false,
    "result": Object {
      "header": Object {
        "connection": "close",
        "content-type": "application/json",
        "date": "Fake",
        "transfer-encoding": "chunked",
      },
      "req": Object {
        "data": undefined,
        "headers": Object {
          "some-header": "1",
          "user-agent": "node-superagent/3.8.3",
        },
        "method": "GET",
        "url": "http://localhost:8080/200?foo",
      },
      "status": 200,
      "text": "{\\"headers\\":{\\"host\\":\\"localhost:8080\\",\\"accept-encoding\\":\\"gzip, deflate\\",\\"user-agent\\":\\"node-superagent/3.8.3\\",\\"some-header\\":\\"1\\",\\"connection\\":\\"close\\"},\\"url\\":\\"/200?foo\\"}",
    },
    "value": Object {
      "headers": Object {
        "accept-encoding": "gzip, deflate",
        "connection": "close",
        "host": "localhost:8080",
        "some-header": "1",
        "user-agent": "node-superagent/3.8.3",
      },
      "url": "/200?foo",
    },
  },
}
`;

exports[`Fetchy fetch GET 404 1`] = `
Object {
  "fetch": [Function],
  "reset": [Function],
  "state": Object {
    "error": null,
    "fulfilled": false,
    "pending": false,
    "progress": null,
    "rejected": false,
    "result": null,
    "value": null,
  },
}
`;

exports[`Fetchy fetch GET 404 2`] = `
Object {
  "result": Object {
    "response": Object {
      "header": Object {
        "connection": "close",
        "date": "Fake",
        "transfer-encoding": "chunked",
      },
      "req": Object {
        "data": undefined,
        "headers": Object {
          "user-agent": "node-superagent/3.8.3",
        },
        "method": "GET",
        "url": "http://localhost:8080/404",
      },
      "status": 404,
      "text": "404",
    },
  },
}
`;

exports[`Fetchy fetch GET 404 3`] = `[Error: Not Found]`;

exports[`Fetchy fetch GET 404 4`] = `
Object {
  "fetch": [Function],
  "reset": [Function],
  "state": Object {
    "error": [Error: Not Found],
    "fulfilled": false,
    "pending": false,
    "progress": null,
    "rejected": true,
    "result": Object {
      "header": Object {
        "connection": "close",
        "date": "Fake",
        "transfer-encoding": "chunked",
      },
      "req": Object {
        "data": undefined,
        "headers": Object {
          "user-agent": "node-superagent/3.8.3",
        },
        "method": "GET",
        "url": "http://localhost:8080/404",
      },
      "status": 404,
      "text": "404",
    },
    "value": null,
  },
}
`;

exports[`Fetchy fetch POST 200 1`] = `
Object {
  "fetch": [Function],
  "reset": [Function],
  "state": Object {
    "error": null,
    "fulfilled": false,
    "pending": false,
    "progress": null,
    "rejected": false,
    "result": null,
    "value": null,
  },
}
`;

exports[`Fetchy fetch POST 200 2`] = `
Object {
  "error": null,
  "fulfilled": true,
  "pending": false,
  "progress": null,
  "rejected": false,
  "result": Object {
    "header": Object {
      "connection": "close",
      "content-length": "0",
      "date": "Fake",
    },
    "req": Object {
      "data": Object {
        "data": Array [
          104,
          101,
          108,
          108,
          111,
          32,
          119,
          111,
          114,
          108,
          100,
          10,
        ],
        "type": "Buffer",
      },
      "headers": Object {
        "user-agent": "node-superagent/3.8.3",
      },
      "method": "GET",
      "url": "http://localhost:8080/post/200",
    },
    "status": 200,
    "text": "",
  },
  "value": Object {},
}
`;

exports[`Fetchy fetch POST 200 3`] = `
Object {
  "fetch": [Function],
  "reset": [Function],
  "state": Object {
    "error": null,
    "fulfilled": true,
    "pending": false,
    "progress": null,
    "rejected": false,
    "result": Object {
      "header": Object {
        "connection": "close",
        "content-length": "0",
        "date": "Fake",
      },
      "req": Object {
        "data": Object {
          "data": Array [
            104,
            101,
            108,
            108,
            111,
            32,
            119,
            111,
            114,
            108,
            100,
            10,
          ],
          "type": "Buffer",
        },
        "headers": Object {
          "user-agent": "node-superagent/3.8.3",
        },
        "method": "GET",
        "url": "http://localhost:8080/post/200",
      },
      "status": 200,
      "text": "",
    },
    "value": Object {},
  },
}
`;
